modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/utils.py

```diff
@@ -250,18 +256,32 @@ def check_cols_to_join(what, df, col_names):
     Tuple[HdkOnNativeDataframe, list]
         The aligned data frame and column names.
     """
-    cols = df.columns
+
+    def check_dtype(name, dtype):
+        if not (
+            is_string_dtype(dtype)
+            or is_any_int_dtype(dtype)
+            or is_datetime64_dtype(dtype)
+            or is_categorical_dtype(dtype)
+        ):
+            raise NotImplementedError(f"Join on column '{name}' of type '{dtype}'")
+
+    dtypes = df.dtypes
     new_col_names = col_names
     for i, col in enumerate(col_names):
-        if col in cols:
+        dtype = dtypes.get(col, None)
+        if dtype is not None:
+            check_dtype(col, dtype)
             continue
+
         new_name = None
         if df._index_cols is not None:
             for c in df._index_cols:
                 if col == df._index_name(c):
                     new_name = c
+                    check_dtype(col, df._dtypes[c])
                     break
-        elif df.has_index_cache:
+        elif df.has_index_cache and df.index.name == col:
```

Comment:
Why? Looks redundant to this pull request.

Reply:
I agree It's not related to this PR. I've changed this line, because it seems to be a bug.